PPM-5 add smart steering persistent data to db

adam1985d opened the pull request:
This PR introduces client parameters to enable smart-steering capabilities and persistency of those parameters across prplmesh restarts and system reboots.

The BPL provides the interface to the persistent DB and the 

The added configuration is saved to the persistent DB via BPL and enables a per-client steering configuration persistent across prplmesh-restarts/reboots.

This PR includes:
Add the client parameters to the Node class (runtime controller DB).
Add setters/getters APIs to the client parameters (get is from runtime DB, set can be to runtime DB or also to persistent DB (write-through))
Add APIs to clear/update client persistent information.
Add API to get all clients from persistent DB

adam1985d committed:
controller: db: node: add persistent client params

PPM-5.
Preparative commit.
Added client persistent parameters to the Node class.
The parameters persistency will be handled by the db class and the BPL
will provide the interface to save/read/clear the client's persistent
parameters.
The client persistent parameters enable the smart-client-steering
feature by enabling client's steering configurations and decisions
persistency across prplmesh restarts/reboots.

Signed-off-by: Adam Dov <adamx.dov@intel.com>

adam1985d committed:
controller: db: add persistent client params set/get stubs

PPM-5.
Preparative commit.
Added setters and getters APIs stubs for the new client persistent
params and clients get, clear, and update APIs.

Signed-off-by: Adam Dov <adamx.dov@intel.com>

adam1985d committed:
controller: db: implement client params set/get in runtime DB

PPM-5.
The operation of getting the client's persistent parameter is by reading
it from the runtime controller DB.
The operation of getting the client's persistent parameter can be to the
runtime controller DB only, or write-through (if the set-persistent-db
is set to true).

Implemented getters from runtime controller DB.
Implemented setters only to runtime controller DB.
(Implementation of set to persistent DB using BPL is provided as a
separate commit)

Signed-off-by: Adam Dov <adamx.dov@intel.com>

vitalybu commented on common/beerocks/bcl/include/bcl/beerocks_defines.h:
```diff
@@ -470,6 +470,8 @@ enum class eChannelScanOperationCode : uint8_t {
     INVALID_PARAMS_CHANNELPOOL
 };
 
+enum class ePersistentParamBool : int8_t { DISABLE = 0, ENABLE = 1, NOT_CONFIGURED = -1 };
```
Please add a doxygen comment explaining this type.
I don't understand its' purpose form the name. Bool suggests a TRUE/FALSE or ENABLED/DISABLED values, but you actually have 3 values.

```suggestion
enum class ePersistentParamBool : int8_t { NOT_CONFIGURED = -1, DISABLED = 0, ENABLED = 1 };
```

morantr replied:
It is not a bool, it is tri-state configuration.
So I would call it eTriStateConf or eConfEnabled.

adam1985d replied:
added Doxygen
reordered the values as suggested

vitalybu commented on controller/src/beerocks/master/db/node.cpp:
```diff
@@ -121,6 +121,25 @@ std::ostream &operator<<(std::ostream &os, const node &n)
+        if (node_type == beerocks::TYPE_CLIENT) {
+            os << "Persistent configuration and data:" << std::endl
+               << "   ClientPersistentDataLastEdit: "
+               << float(std::chrono::duration_cast<std::chrono::duration<double>>(
```
Why it this a double casted to float? In what units is the duration?

adam1985d replied:
I changed the way I print this information - please re-review it and let me know what you think

vitalybu commented on controller/src/beerocks/master/db/node.cpp:
```diff
@@ -121,6 +121,25 @@ std::ostream &operator<<(std::ostream &os, const node &n)
+                            .count())
+               << std::endl
+               << "   ClientTimeLifeDelay: "
+               << int(n.client_time_life_delay_sec.count() / (60 * 60 * 24)) << " days" << std::endl
```
```suggestion
               << int(std::chrono::duration_cast<std::chrono::days>(n.client_time_life_delay_sec).count()) << " days" << std::endl
```

adam1985d replied:
`std::chrono::days` isn't available - but will use the `std::chrono::hours` and `/24` only

vitalybu commented on controller/src/beerocks/master/db/node.h:
```diff
@@ -267,6 +267,33 @@ class node {
+    std::chrono::steady_clock::time_point client_persistent_data_last_edit =
+        std::chrono::steady_clock::time_point::min();
+    // optional - if configured the client has its own configured timelife delay
+    std::chrono::seconds client_time_life_delay_sec = std::chrono::seconds(0);
```
```suggestion
    std::chrono::seconds client_time_life_delay_sec = std::chrono::seconds::zero();
```

adam1985d replied:
replaced all locations of `std::chrono::seconds(0)` to use `std::chrono::seconds::zero()`

vitalybu commented on controller/src/beerocks/master/db/db.cpp:
```diff
@@ -2418,6 +2418,80 @@ const std::list<sChannelScanResults> &db::get_channel_scan_results(const sMacAdd
+    return beerocks::ePersistentParamBool::NOT_CONFIGURED;
+}
+
+bool db::set_client_initial_radio(const sMacAddr &mac, const sMacAddr &initial_radio_mac,
```
Why do we need `set_client_stay_on_initial_radio` and `set_client_initial_radio`?

adam1985d replied:
the client can configure a client to stay on its initial radio even before the client was connected for the first time, so the information should be able to be configured separately.
we might also want to save in the future the `client_initial_radio` even if the `stay_on_initial_radio` is false

vitalybu commented on controller/src/beerocks/master/db/db.cpp:
```diff
@@ -2418,6 +2418,80 @@ const std::list<sChannelScanResults> &db::get_channel_scan_results(const sMacAdd
+
+bool db::set_client_time_life_delay(const sMacAddr &mac,
+                                    const std::chrono::seconds &time_life_delay_sec,
+                                    bool set_persistent_db)
```
What is `set_persistent_db` for?

vitalybu commented on controller/src/beerocks/master/db/db.h:
```diff
@@ -589,6 +589,158 @@ class db {
+    std::chrono::steady_clock::time_point get_client_persistent_data_last_edit(const sMacAddr &mac);
+
+    /**
+     * @brief Set the client's time-life delay.
```
What is "time-life"?

adam1985d replied:
added a description to the parameter (in node) and elaborate a bit more in the function description

vitalybu commented on controller/src/beerocks/master/db/db.h:
```diff
@@ -589,6 +589,158 @@ class db {
+    std::chrono::seconds get_client_time_life_delay(const sMacAddr &mac);
+
+    /**
+     * @brief Set the client's stay-on-initial-radio.
```
What is "stay-on-initial-radio"?
I think you need to add a short explanation in the comment about these features..

adam1985d replied:
added a description to the parameter (in node) and elaborate a bit more in the function description

vitalybu commented on controller/src/beerocks/master/db/db.h:
```diff
@@ -589,6 +589,158 @@ class db {
+     * @return true on success.
+     * @return false on failure.
+     */
+    bool set_client_stay_on_initial_radio(const sMacAddr &mac, bool stay_on_initial_radio,
```
```suggestion
    bool set_force_client_to_initial_radio(const sMacAddr &mac, bool force_to_initial_radio,
```

vitalybu commented on controller/src/beerocks/master/db/db.h:
```diff
@@ -589,6 +589,158 @@ class db {
+     * @return true on success.
+     * @return false on failure.
+     */
+    bool set_client_initial_radio(const sMacAddr &mac, const sMacAddr &initial_radio_mac,
```
Can you really set the radio before the client connected?
Or is the purpose of this method is to update the MAC of the radio the client connected to?

adam1985d replied:
this API is an interface for the controller/tasks.
the `set_stay_on_initial_radio` on the other hand can be triggered by the son_management as a response to bml-request

vitalybu commented on controller/src/beerocks/master/db/db.h:
```diff
@@ -589,6 +589,158 @@ class db {
+     * 
+     * @param mac MAC address of client.
+     * @param initial_radio_mac the mac of radio that the client has initially connected to.
+     * @param set_persistent_db set-persistent-db flag - default is true (write-through).
```
You can't describe a variable with the variable's name but changing underscores to hypens :)

adam1985d replied:
true.
changed to `* @param set_persistent_db if set to true, update the persistent-db (write-through), default is true.`

vitalybu commented on controller/src/beerocks/master/db/db.h:
```diff
@@ -589,6 +589,158 @@ class db {
+     * @return an empty map of clients is returned if the persistent db is empty
+     */
+    std::unordered_map<std::string, std::unordered_map<std::string, std::string>>
+    get_all_persistent_db_clients();
```
```suggestion
    load_persistent_db();
```
?

adam1985d replied:
`load_persistent_db_clients`
?

vitalybu replied:
Why do you want to make this method client-specific?
Tomorrow we can additional persistent values, why not have a single method for loading all the information from the DB?

vitalybu commented on controller/src/beerocks/master/db/db.cpp:
```diff
@@ -2423,75 +2423,252 @@ const std::list<sChannelScanResults> &db::get_channel_scan_results(const sMacAdd
+    auto timestamp = std::chrono::steady_clock::now();
+    if (set_persistent_db) {
+        LOG(DEBUG) << __FUNCTION__ << ", configuring persistent-db, stay_on_initial_radio = "
+                   << string_utils::bool_str(stay_on_initial_radio);
```
```suggestion
                   << std::boolalpha << stay_on_initial_radio;
```

adam1985d replied:
wow,
we have so many places in code using the `string_utils::bool_str`
👍

vitalybu commented on controller/src/beerocks/master/db/db.cpp:
```diff
@@ -2423,75 +2423,252 @@ const std::list<sChannelScanResults> &db::get_channel_scan_results(const sMacAdd
+    }
+
+    LOG(DEBUG) << __FUNCTION__
+               << ", stay_on_initial_radio = " << tlvf::mac_to_string(initial_radio_mac);
```
```suggestion
               << ", stay_on_initial_radio = " << initial_radio_str;
```

adam1985d replied:
fixed

morantr commented on common/beerocks/bcl/include/bcl/beerocks_defines.h:
```diff
@@ -470,6 +470,8 @@ enum class eChannelScanOperationCode : uint8_t {
     INVALID_PARAMS_CHANNELPOOL
 };
 
+enum class ePersistentParamBool : int8_t { DISABLE = 0, ENABLE = 1, NOT_CONFIGURED = -1 };
```
I don't think it should be defined here. It is used only by the controller, its database to be accurate. 
I think it should be defined there, and not on beerocks_defines.h which almost all files on the project includes.

adam1985d replied:
moved it to the `node.h`

morantr commented on controller/src/beerocks/master/db/node.cpp:
```diff
@@ -121,6 +121,29 @@ std::ostream &operator<<(std::ostream &os, const node &n)
+            std::chrono::hours client_persistent_data_last_edit_hours =
+                std::chrono::duration_cast<std::chrono::hours>(tCurrTime -
+                                                               n.client_persistent_data_last_edit);
+            std::chrono::hours client_time_life_delay_hours =
```
Since this integers represent the time on hours, I would call the `count()` directly from here, and also use `auto` decleration:
```cpp
auto client_persistent_data_last_edit_hours = std::chrono::duration_cast<std::chrono::hours>(tCurrTime -n.client_persistent_data_last_edit).count();

auto client_time_life_delay_hours =  std::chrono::duration_cast<std::chrono::hours>(n.client_time_life_delay_sec).count();

```

adam1985d replied:
fixed

morantr commented on controller/src/beerocks/master/db/node.cpp:
```diff
@@ -121,6 +121,29 @@ std::ostream &operator<<(std::ostream &os, const node &n)
+               << int(client_persistent_data_last_edit_hours.count() / 24) << " days, "
+               << int(client_persistent_data_last_edit_hours.count() % 24) << " hours" << std::endl
+               << "   ClientTimeLifeDelay: " << int(client_time_life_delay_hours.count() / 24)
+               << " days, " << int(client_time_life_delay_hours.count() % 24) << " hours"
```
The `count()` function should return printable return value, so no need to cast it to int.

adam1985d replied:
removed the cast

morantr commented on controller/src/beerocks/master/db/node.cpp:
```diff
@@ -121,6 +121,29 @@ std::ostream &operator<<(std::ostream &os, const node &n)
+               << "   ClientInitialRadio: " << n.client_initial_radio << std::endl
+               << "   ClientStayOnSelectedBand: " << int(n.client_stay_on_selected_band)
+               << std::endl
+               << "   ClientSelectedBands: " << int(n.client_selected_bands) << std::endl;
```
Suggestion:
Add an operator overload for `<<` so when using it on stream, it will print "Enabled", "Disabled" or "Not Configured".
(nitpick)

adam1985d replied:
done

morantr commented on controller/src/beerocks/master/db/node.h:
```diff
@@ -267,6 +267,33 @@ class node {
+
+    /*
+     * persistent configurations - end
+     */
```
It would look better if each comment would start with a capital letter and end with a dot.
Also, a space between declaration to next declaration comment would look nicer IMO:
```cpp
Foo foo;

// Bar comment.
Bar bar;
```
(nitpick)

adam1985d replied:
fixed - if that makes it cleared for other devs I'm happy to adopt

morantr commented on controller/src/beerocks/master/db/db.h:
```diff
@@ -589,6 +589,158 @@ class db {
+    /**
+     * @brief Get the client's last persistent data edit time.
+     * 
+     * @param mac MAC address of client.
```
"... of **a** client".
This needs to be fixed on several places on this file.

adam1985d replied:
done

morantr commented on controller/src/beerocks/master/db/db.h:
```diff
@@ -589,6 +589,158 @@ class db {
+     * @brief Set the client's time-life delay.
+     * 
+     * @param mac MAC address of client.
+     * @param time_life_delay_sec client-specific aging time.
```
Please make sure to start the argument description with a capital letter.

adam1985d replied:
fixed all occurrences

morantr commented on controller/src/beerocks/master/db/db.h:
```diff
@@ -589,6 +589,158 @@ class db {
+     * @param set_persistent_db if set to true, update the persistent-db (write-through), default is true.
+     * 
+     * @return true on success.
+     * @return false on failure.
```
@return true on success, otherwise false.

adam1985d replied:
replaced all occurrences

morantr commented on controller/src/beerocks/master/db/db.h:
```diff
@@ -589,6 +589,158 @@ class db {
+     * @param mac MAC address of client.
+     * @param time_life_delay_sec client-specific aging time.
+     * @param set_persistent_db if set to true, update the persistent-db (write-through), default is true.
+     * 
```
No need to add a space between @param and @return. Only after the @brief.

adam1985d replied:
ok - removed extra spacing

simonbaren commented on controller/src/beerocks/master/db/db.cpp:
```diff
@@ -2423,73 +2423,257 @@ const std::list<sChannelScanResults> &db::get_channel_scan_results(const sMacAdd
+    }
+
+    n->client_time_life_delay_sec       = time_life_delay_sec;
+    n->client_persistent_data_last_edit = timestamp;
```
AFAIU - the naming here is misleading `client_persistent_data_last_edit` - in the case that `set_persistent_db` is set to false you are not actually updating the persistent DB, you only update the controller runtime DB.

morantr replied:
I also don't understand. I thought that if `set_persistent_db` is set, we should save something on the persistent db.

adam1985d replied:
@simonbaren 
client persistent data can be set to the runtime db with multiple calls and then a call to `update_client_persistent_db()` will sync the runtime configuration to the persistent DB. the timestamp is handling the time of last edit (of configuration) and not the last time of actual synchronization to the persistent DB.
in any case - as part of the orderly shutdown of the controller - the runtime configuration should be saved to the persistent db - but again - with the timestamp indicating the last time of client configuration edit and not when it was synched to the persistent DB!
and naming suggestion is welcomed

@morantr - this is correct, but I will quote the draft PR description:
"Draft PR - missing the calls to BPL APIs will be added after #1460 is merged"
(at time of the draft PR creation BPL interface was not available yet)

morantr commented on controller/src/beerocks/master/db/db.h:
```diff
@@ -589,6 +589,158 @@ class db {
+     * 
+     * @param mac MAC address of client.
+     * @param selected_bands: client selected band/bands. FREQ_UNKNOWN is considered as "not-configured"
+     * @param set_persistent_db: set-persistent-db flag - default is true (write-through).
```
You should not add colons between the argument name to the description.
Second, does set_persistent_db means "Save configuration on persistent database"?
Because it is much more understandable (at least for me) to phrase it like that.

adam1985d replied:
done

morantr commented on controller/src/beerocks/master/db/db.h:
```diff
@@ -589,6 +589,158 @@ class db {
+     * @return an empty map of clients is returned if the persistent db is empty
+     */
+    std::unordered_map<std::string, std::unordered_map<std::string, std::string>>
+    get_all_persistent_db_clients();
```
Please add dots at the end of the sentences.
Second, maybe you could define a type for this return value with Doxygen documentation and then use it here. This could make the return value explanation much more simple.
Also, when using MAC address please use sMacAddr and not std::string.

adam1985d replied:
done

morantr commented on controller/src/beerocks/master/db/db.cpp:
```diff
@@ -2423,73 +2423,257 @@ const std::list<sChannelScanResults> &db::get_channel_scan_results(const sMacAdd
 {
+    auto n = get_node(mac);
+    if (!n) {
+        LOG(ERROR) << "node not found.... ";
```
What are the dots at the end of the sentence are good for? 
Also, I think that printing the given node is valuable information.

Apply this to several places on this file.

adam1985d replied:
fixed

morantr commented on controller/src/beerocks/master/db/db.cpp:
```diff
@@ -2423,73 +2423,257 @@ const std::list<sChannelScanResults> &db::get_channel_scan_results(const sMacAdd
+        LOG(ERROR) << "node not found.... ";
+        return false;
+    } else if (n->get_type() != beerocks::TYPE_CLIENT) {
+        LOG(ERROR) << __FUNCTION__ << "node " << tlvf::mac_to_string(mac)
```
No need to ` tlvf::mac_to_string(mac)`, it is done automatically.

adam1985d replied:
removed it

morantr commented on controller/src/beerocks/master/db/db.cpp:
```diff
@@ -2423,73 +2423,257 @@ const std::list<sChannelScanResults> &db::get_channel_scan_results(const sMacAdd
+        return false;
+    }
+
+    LOG(DEBUG) << __FUNCTION__ << ", time_life_delay_sec = " << time_life_delay_sec.count();
```
If time_life_delay_sec is in a unit of seconds, Why do you use `count()` on it? It doesn't make sense.
I would expect that it will be really with a second unit.

adam1985d replied:
as we dicsussed offline, leaving it as is since the main usage is in the std::chrono:seconds() type

morantr commented on controller/src/beerocks/master/db/db.cpp:
```diff
@@ -2423,73 +2423,257 @@ const std::list<sChannelScanResults> &db::get_channel_scan_results(const sMacAdd
+    auto timestamp = std::chrono::steady_clock::now();
+    if (set_persistent_db) {
+        LOG(DEBUG) << __FUNCTION__
+                   << ", configuring persistent-db, stay_on_initial_radio = " << std::boolalpha
```
boolalpha is configured by default on beerocks logging. No need to add it.

adam1985d replied:
reomved it

morantr reviewed (approved):
Nothing critical, so approving.

adam1985d committed:
controller: db: add static helper functions

PPM-5.
Preparative commit.
added helper functions for translation of persistent client
configurations to/from string.
added helper functions for translation of db_entry of client to/from mac
address.

Signed-off-by: Adam Dov <adamx.dov@intel.com>

adam1985d committed:
controller: db: add private function to update client entry

PPM-5.
Preparative commit.
Added a private function that encapsulates any required validation
before updating or adding the client entry in the persistent db using
the BPL APIs.

Signed-off-by: Adam Dov <adamx.dov@intel.com>

adam1985d committed:
controller: db: save client persistent params to bpl db

PPM-5.
Implemented saving client persistent data to the persistent DB using the
"bpl::db_" APIs.
Implemented function to check if the client exists in persistent DB.
Implemented add/remove of client entry to/from persistent DB.
Implemented a function to save all client's configured persistent data
from runtime controller DB to persistent controller DB.

Signed-off-by: Adam Dov <adamx.dov@intel.com>

vitalybu commented on controller/src/beerocks/master/db/node.h:
```diff
@@ -35,6 +35,14 @@ typedef struct {
+ * @brief Extended boolean parameter to support "not configured" value for persistent configuration.
+ * For persistent data, it is important to differ between configured enable/disable to uncofigured value.
+ */
+enum class ePersistentParamBool : int8_t { NOT_CONFIGURED = -1, DISABLE = 0, ENABLE = 1 };
```
```suggestion
enum class ePersistency : int8_t { NOT_CONFIGURED = -1, DISABLE = 0, ENABLE = 1 };
```

adam1985d replied:
the enum doesn't only provide an indication of whether the value is persistent or not - it also provides its value.
I am open to naming suggestions but think `ePersistency` is misleading.

vitalybu commented on controller/src/beerocks/master/db/db.cpp:
```diff
@@ -4018,3 +4019,22 @@ void db::set_prplmesh(const sMacAddr &mac)
+{
+    auto db_entry = client_db_entry_from_mac(mac);
+
+    if (!bpl::db_has_entry(type_to_string(beerocks::eType::TYPE_CLIENT), db_entry)) {
```
If you convert the type to string more than once, it's probably better to use a variable.

adam1985d replied:
fixed on all occurrences

vitalybu commented on controller/src/beerocks/master/db/db.cpp:
```diff
@@ -2475,11 +2475,38 @@ const std::list<sChannelScanResults> &db::get_channel_scan_results(const sMacAdd
 {
+    auto db_entry = client_db_entry_from_mac(mac);
+
+    if (!bpl::db_has_entry(std::string("type"), db_entry)) {
```
Why not simply:
```suggestion
    if (!bpl::db_has_entry("type", db_entry)) {
```

vitalybu replied:
Also, what is the API of `bpl::db_has_entry()`?
What does it mean to `bpl::db_has_entry("type", ...)`?

adam1985d replied:
an entry (in UCI correlates to section) has a type and a name - all clients have a section-type `client` and looking them up in DB is by the client mac which is used as its name.
similar to the `db::has_node()` the `bpl::db_has_entry()` returns true if entry exist in DB.
to prevent a situation where a mac of the same client was used to describe something that **is not** of section-type `client` we can also specify expected entry type. (if we do not provide it - no validation of entry-type will be performed)

```
/**
 * @brief find if entry exists
 * 
 * @param[in] entry_type type of the requested entry.
 * @param[in] entry_name name of the requested entry.
 * @return true if entry exists in db, false otherwise
 */
bool db_has_entry(const std::string &entry_type, const std::string &entry_name);
```

vitalybu commented on controller/src/beerocks/master/db/db.cpp:
```diff
@@ -2516,6 +2543,16 @@ bool db::set_client_time_life_delay(const sMacAddr &mac,
                    << ", configuring persistent-db, timelife = " << time_life_delay_sec.count();
+
+        std::unordered_map<std::string, std::string> values_map;
+        values_map.at("timestamp") = timestamp_to_string_seconds(timestamp);
```
Why do you need a function to convert the timestamp to seconds?

adam1985d replied:
I am using this conversion in many places in the code so I thought it would be beneficial to have:
1. meaningful function to understand the operation performed
2. single-point-of-change in case we wish to change how we save the timestamp in the future

vitalybu commented on controller/src/beerocks/master/db/db.cpp:
```diff
@@ -2721,16 +2798,102 @@ bool db::clear_client_persistent_db(const sMacAddr &mac)
+    std::unordered_map<std::string, std::string> values_map;
+
+    //fill values map of client persistent params
+    values_map.at("timestamp") = timestamp_to_string_seconds(n->client_parameters_last_edit);
```
All static strings (e.g. "timestamp", "timelife" etc.) should be constexpr.

adam1985d replied:
the map expects `std::string` while in c++11 it is not supported as `constexpr`
I have changed to `static const std::string` at the beginning of the `db.cpp` file

reference to the mentioned above:
https://stackoverflow.com/questions/27123306/is-it-possible-to-use-stdstring-in-a-constexpr

P.S. - I didn't use the `constexpr char constString[] = "constString";` example they suggested as the this would cause a runtime `std::string` creation from `char[]` on every use.

VladyslavTupikin commented on controller/src/beerocks/master/db/db.cpp:
```diff
@@ -19,6 +20,66 @@ using namespace beerocks_message;
+// static
+std::string db::type_to_string(beerocks::eType type)
+{
+    if (type == beerocks::eType::TYPE_GW) {
```
I think the switch-case will be better than if-else here.

adam1985d replied:
both options are valid.
I actually used to use switch-case for such implementations but received comments exactly the opposite to your comment ("please use if-else") - I guess it is a matter of taste

VladyslavTupikin replied:
I disagree with you. The switch-case mechanism was written especially for situations as we can see here and use if-else for replacing standard switch-case looks weird.

MuteSpirit replied:
I vote to `switch-case` only because the shorter code

adam1985d replied:
replaced with switch-case that return `"*"` instead of `std::string("*")`

VladyslavTupikin commented on controller/src/beerocks/master/db/db.cpp:
```diff
@@ -2418,6 +2479,431 @@ const std::list<sChannelScanResults> &db::get_channel_scan_results(const sMacAdd
+                                    const std::chrono::seconds &time_life_delay_sec,
+                                    bool save_to_persistent_db)
+{
+    auto n = get_node(mac);
```
```suggestion
    auto node = get_node(mac);
```

adam1985d replied:
replaced all new-code occurrences of `n` to `node`

VladyslavTupikin commented on controller/src/beerocks/master/db/db.cpp:
```diff
@@ -2418,6 +2479,431 @@ const std::list<sChannelScanResults> &db::get_channel_scan_results(const sMacAdd
+
+std::chrono::seconds db::get_client_time_life_delay(const sMacAddr &mac)
+{
+    auto n = get_node(mac);
```
```suggestion
    auto node = get_node(mac);
```

adam1985d replied:
replaced all new-code occurrences of `n` to `node`

VladyslavTupikin commented on controller/src/beerocks/master/db/db.cpp:
```diff
@@ -2418,6 +2479,431 @@ const std::list<sChannelScanResults> &db::get_channel_scan_results(const sMacAdd
+bool db::set_client_stay_on_initial_radio(const sMacAddr &mac, bool stay_on_initial_radio,
+                                          bool save_to_persistent_db)
+{
+    auto n = get_node(mac);
```
```suggestion
    auto node = get_node(mac);
```

adam1985d replied:
replaced all new-code occurrences of `n` to `node`

VladyslavTupikin commented on controller/src/beerocks/master/db/db.cpp:
```diff
@@ -2418,6 +2479,431 @@ const std::list<sChannelScanResults> &db::get_channel_scan_results(const sMacAdd
+    LOG(DEBUG) << __FUNCTION__ << ", stay_on_initial_radio = " << stay_on_initial_radio;
+
+    auto timestamp = std::chrono::steady_clock::now();
+    if (save_to_persistent_db) {
```
This code is copy-paste from the method `set_client_time_life_delay`. I think you should create a new method for updating the database for avoid copy-paste issues.

adam1985d replied:
the next commit introduces the core-functionality which is inside this condition and moving it to external function more than it is in the next commit is not really beneficial IMO

VladyslavTupikin commented on controller/src/beerocks/master/db/db.cpp:
```diff
@@ -2418,6 +2479,431 @@ const std::list<sChannelScanResults> &db::get_channel_scan_results(const sMacAdd
+
+ePersistentParamBool db::get_client_stay_on_initial_radio(const sMacAddr &mac)
+{
+    auto n = get_node(mac);
```
```suggestion
    auto node = get_node(mac);
```

adam1985d replied:
replaced all new-code occurrences of `n` to `node`

VladyslavTupikin commented on controller/src/beerocks/master/db/db.cpp:
```diff
@@ -2418,6 +2479,431 @@ const std::list<sChannelScanResults> &db::get_channel_scan_results(const sMacAdd
+bool db::set_client_initial_radio(const sMacAddr &mac, const sMacAddr &initial_radio_mac,
+                                  bool save_to_persistent_db)
+{
+    auto n = get_node(mac);
```
```suggestion
    auto node = get_node(mac);
```

adam1985d replied:
replaced all new-code occurrences of `n` to `node`

VladyslavTupikin commented on controller/src/beerocks/master/db/db.cpp:
```diff
@@ -2418,6 +2479,431 @@ const std::list<sChannelScanResults> &db::get_channel_scan_results(const sMacAdd
+    LOG(DEBUG) << __FUNCTION__ << ", initial_radio = " << initial_radio_mac;
+
+    auto timestamp = std::chrono::steady_clock::now();
+    if (save_to_persistent_db) {
```
Code copy-paste. See the commentary above.

adam1985d replied:
see my response above

VladyslavTupikin commented on controller/src/beerocks/master/db/db.cpp:
```diff
@@ -2418,6 +2479,431 @@ const std::list<sChannelScanResults> &db::get_channel_scan_results(const sMacAdd
+
+sMacAddr db::get_client_initial_radio(const sMacAddr &mac)
+{
+    auto n = get_node(mac);
```
```suggestion
    auto node = get_node(mac);
```

adam1985d replied:
replaced all new-code occurrences of `n` to `node`

VladyslavTupikin commented on controller/src/beerocks/master/db/db.cpp:
```diff
@@ -2418,6 +2479,431 @@ const std::list<sChannelScanResults> &db::get_channel_scan_results(const sMacAdd
+bool db::set_client_stay_on_selected_band(const sMacAddr &mac, bool stay_on_selected_band,
+                                          bool save_to_persistent_db)
+{
+    auto n = get_node(mac);
```
```suggestion
    auto node = get_node(mac);
```

adam1985d replied:
replaced all new-code occurrences of `n` to `node`

VladyslavTupikin commented on controller/src/beerocks/master/db/db.cpp:
```diff
@@ -2418,6 +2479,431 @@ const std::list<sChannelScanResults> &db::get_channel_scan_results(const sMacAdd
+    LOG(DEBUG) << __FUNCTION__ << ", stay_on_selected_band = " << stay_on_selected_band;
+
+    auto timestamp = std::chrono::steady_clock::now();
+    if (save_to_persistent_db) {
```
Copy-paste code.

adam1985d replied:
see my comment above

VladyslavTupikin commented on controller/src/beerocks/master/db/db.cpp:
```diff
@@ -2418,6 +2479,431 @@ const std::list<sChannelScanResults> &db::get_channel_scan_results(const sMacAdd
+
+ePersistentParamBool db::get_client_stay_on_selected_band(const sMacAddr &mac)
+{
+    auto n = get_node(mac);
```
```suggestion
    auto node = get_node(mac);
```

adam1985d replied:
replaced all new-code occurrences of `n` to `node`

VladyslavTupikin commented on controller/src/beerocks/master/db/db.cpp:
```diff
@@ -2418,6 +2479,431 @@ const std::list<sChannelScanResults> &db::get_channel_scan_results(const sMacAdd
+bool db::set_client_selected_bands(const sMacAddr &mac, beerocks::eFreqType selected_bands,
+                                   bool save_to_persistent_db)
+{
+    auto n = get_node(mac);
```
```suggestion
    auto node = get_node(mac);
```

adam1985d replied:
replaced all new-code occurrences of `n` to `node`

VladyslavTupikin commented on controller/src/beerocks/master/db/db.cpp:
```diff
@@ -2418,6 +2479,431 @@ const std::list<sChannelScanResults> &db::get_channel_scan_results(const sMacAdd
+    LOG(DEBUG) << __FUNCTION__ << ", selected_band = " << int(selected_bands);
+
+    auto timestamp = std::chrono::steady_clock::now();
+    if (save_to_persistent_db) {
```
Copy-paste code.

adam1985d replied:
see my comments above

VladyslavTupikin commented on controller/src/beerocks/master/db/db.cpp:
```diff
@@ -2418,6 +2479,431 @@ const std::list<sChannelScanResults> &db::get_channel_scan_results(const sMacAdd
+
+beerocks::eFreqType db::get_client_selected_bands(const sMacAddr &mac)
+{
+    auto n = get_node(mac);
```
```suggestion
    auto node = get_node(mac);
```

adam1985d replied:
replaced all new-code occurrences of `n` to `node`

VladyslavTupikin commented on controller/src/beerocks/master/db/db.cpp:
```diff
@@ -2418,6 +2479,431 @@ const std::list<sChannelScanResults> &db::get_channel_scan_results(const sMacAdd
+
+bool db::clear_client_persistent_db(const sMacAddr &mac)
+{
+    auto n = get_node(mac);
```
```suggestion
    auto node = get_node(mac);
```

adam1985d replied:
replaced all new-code occurrences of `n` to `node`

VladyslavTupikin commented on controller/src/beerocks/master/db/db.cpp:
```diff
@@ -2418,6 +2479,431 @@ const std::list<sChannelScanResults> &db::get_channel_scan_results(const sMacAdd
+
+bool db::update_client_persistent_db(const sMacAddr &mac)
+{
+    auto n = get_node(mac);
```
```suggestion
    auto node = get_node(mac);
```

adam1985d replied:
replaced all new-code occurrences of `n` to `node`

VladyslavTupikin commented on controller/src/beerocks/master/db/node.cpp:
```diff
@@ -29,6 +29,18 @@ node::node(beerocks::eType type_, const std::string &mac_)
+    return os;
+}
+
 std::ostream &operator<<(std::ostream &os, const node &n)
```
```suggestion
std::ostream &operator<<(std::ostream &os, const node &node)
```
Looks weird but in the code, I prefer to see node instead of magic n.

adam1985d replied:
on this specific location I am going to leave as is - but on the new introduced APIs (in following commits) will change from `n` to `node`

VladyslavTupikin replied:
agree

VladyslavTupikin commented on controller/src/beerocks/master/db/db.cpp:
```diff
@@ -19,6 +20,66 @@ using namespace beerocks_message;
+std::string db::client_db_entry_from_mac(const sMacAddr &mac)
+{
+    std::string db_entry = tlvf::mac_to_string(mac);
+    std::replace(db_entry.begin(), db_entry.end(), ':', '_');
```
Why do we need to replace ':' character?

adam1985d replied:
uci doesn't support `:` in the section name

VladyslavTupikin commented on controller/src/beerocks/master/db/db.cpp:
```diff
@@ -2418,6 +2479,431 @@ const std::list<sChannelScanResults> &db::get_channel_scan_results(const sMacAdd
+
+std::chrono::steady_clock::time_point db::get_client_parameters_last_edit(const sMacAddr &mac)
+{
+    auto n = get_node(mac);
```
```suggestion
    auto node = get_node(mac);
```

adam1985d replied:
replaced all new-code occurrences of `n` to `node`

MuteSpirit commented on controller/src/beerocks/master/db/db.cpp:
```diff
@@ -19,6 +20,66 @@ using namespace beerocks_message;
+std::string db::type_to_string(beerocks::eType type)
+{
+    if (type == beerocks::eType::TYPE_GW) {
+        return std::string("gateway");
```
Why not use
```
return "gateway";
```
and not explicitly create `std::string` object?

adam1985d replied:
replaced with switch-case that return `"*"` instead of `std::string("*")`
